exercises/dnd-character/src/test/java/DnDCharacterTest.java

```diff
@@ -128,7 +130,144 @@ public void testRandomCharacterIsValid() {
+            assertEquals(initialConstitution, dndCharacter.getConstitution());
+            assertEquals(initialIntelligence, dndCharacter.getIntelligence());
+            assertEquals(initialWisdom,       dndCharacter.getWisdom());
+            assertEquals(initialCharisma,     dndCharacter.getCharisma());
+        }
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void testAbilityIsProbablyNotConstant() {
+        int ability = dndCharacter.ability();
+        for (int i = 0; i < 100; i++) {
+            if (ability != dndCharacter.ability()) {
+                return;
+            }
+        }
+        fail();
     }
 
+    @Ignore("Remove to run test")
+    @Test
+    public void testAbilityIsDistributedAccordingToRules() {
+        /* This test is meant to check if the 'ability' method is actually
+         * returning a (pseudo-)random number which respects the rules of the
+         * game:
+         *     Roll four dice. Take the sum of the largest three dice.
+         *
+         * This test was added to filter out two common, wrong implementations:
+         *     Returning a random number between 3 and 18.
+         *     Returning the sum of three random numbers each between 1 and 6.
+         *
+         * The way we check if the 'ability' method is distributed correctly is
+         * by creating a histogram of the distribution by calling ability
+         * repeatedly. Then we compare it to the reference distribution within
+         * a certain tolerance.
+         */
+
+        // These two values were chosen empirically.
+        final int cycles       = 10000;
+        final double tolerance = 0.02;
+
+        // Amount of possible combinations for throwing 4 dice
+        int referenceCount = 6 * 6 * 6 * 6;
+        // This distribution was generated by going through all possible
+        // combinations for throwing four dice, computing the ability for that
+        // combination, and incrementing the array at that index.
+        int[] referenceDistribution = new int[] {
+                0, 0, 0, 1, 4, 10, 21, 38, 62, 91, 122, 148, 167, 172, 160, 131, 94, 54, 21
+        };
+        int[] actualDistribution = new int[19];
+
+        for (int i = 0; i < cycles; i++) {
+            int ability = dndCharacter.ability();
+            if (ability >= 3 && ability <= 18) {
+                actualDistribution[ability]++;
+            } else {
+                fail("ability out of Range");
+            }
+        }
+
+        for (int ability = 3; ability <= 18; ability++) {
+            double referenceProbability = (double) referenceDistribution[ability] / referenceCount;
+            double actualProbability    = (double) actualDistribution[ability] / cycles;
+            double difference           = Math.abs(referenceProbability - actualProbability);
+            if (difference > tolerance) {
+                fail();
+            }
+        }
+    }
+
+    /*
+     * Extra Notes:
```

Comment:
This is a very interesting statistical commentary, but I'm not sure it belongs here. Perhaps this sort of thing should go in the mentoring notes for this exercise?

Reply:
Yes, this really remains to be decided.

The purpose of the exercise (I can say because I came up with it) is to explore probabilistic property testing.

But any maintainer is fit to choose how to make any exercise fit some track-specific goal.

I think this stuff is ingenious -- and it is, it seems, one of the things people do wrong: Not understanding that just because the range is correct, the distribution isn't necessarily.

Alternative suggestions for placements:

 - In a hint: Would increase exposure.
 - In a git commit message: Would remove exposure.
 - In a blog post linked to from a hint: Would require someone to want to write a blog post.

Reply:
The idea for this comment was, that if a student has a wrong solution which fails the `testAbilityIsDistributedAccordingToRules` test, but they don't jet know why it is failing, they might want to check out the source of the tests and thus stumble on this note. I admit that probably not every student will find it, but for some it might help them solve the exercise.

Reply:
Personally, I am leaning towards having a hints.md file that includes this extra information about this test. If it makes you feel better, put a Javadoc comment on the test method telling them to check the hints.